RELEASE_NOTES.md

```diff
@@ -0,0 +1,69 @@
+# 3.4 
+
+The top four improvements in Dafny 3.4 are:
+- For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program.
+- Add command line options to assist in debugging verification performance.
+- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 
+- The C# back-end supports traits as type parameters on datatypes.
+
+### Verification
+- fix: Resolve unsoundness issue (https://github.com/dafny-lang/dafny/issues/1619).
+- fix: Don't silently succeed if the solver crashes (https://github.com/boogie-org/boogie/pull/488).
+- feat: Prevent changes in the verification behavior of a proof, when any of these types of changes are made to Dafny user code:
+  - Changes logically unrelated to the proof being verified
+  - Changes to the name of any declaration
+  - Changes to the order of top-level declarations
+- feat: Assist in debugging the verification performance of a proof by adding the `/vcsSplitOnEveryAssert` CLI option and `{:vcs_split_on_every_assert}` attribute, and report the outcome and duration of splits when they occur in `/verificationLogger:trx` content.
+- feat: Add a `/verificationLogger:csv` CLI option that emits the same status and timing information as `/verificationLogger:trx`, but in an easier-to-parse format, along with Z3 resource counts for more repeatable tracking of verification difficulty.
+
+### IDE 
+- fix: Clean up process resources if IDE closed or restarted.
+- fix: Do not let the Dafny compilation status bar get in a stuck state.
+
+- feat: Verification status reporting shows which proof is being verified, which can help debug slow to verify proofs.
+- feat: Publish parsing and resolution diagnostics before verification has completed. Verification diagnostics from previous runs are migrated.
+- feat: Enable 'go to definition', 'hover' and 'signature help' features before verification has completed.
+- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
+- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.
+- feat: Display related resolution errors using nested diagnostics instead of independent diagnostics.
+
+### UX
+- feat: Improve error reporting when providing incorrectly typed arguments in a function call.
+- feat: Improve error reporting when using type tests.
+
+### C#
+- fix: resolve various instances where Dafny would produce invalid C# code (#1607, #1761, and #1762).
+- feat: Support variant type parameters on datatype definitions, which enables using traits as type arguments (https://github.com/dafny-lang/dafny/issues/1499).
+- feat: Support for downcasting both custom datatypes and functions.
```

Comment:
Could we be more specific and/or link the issue here? The description is a little ambiguous - does it mean explicit `is` tests, or implicit downcasting because the verifier proved no `is` is necessary?

Reply:
Probably for @fabiomadge to answer?

Reply:
This should affect both explicit `as` casts and implicit ones. A runtime check using `is` is unrelated.